_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,791 @@
+#[derive(sqlx::FromRow, Serialize)]
+pub struct WebsiteStats {
+    time: DateTime<Utc>,
+    uptime_pct: Option<i16>,
+}
+
+// holds website url/alias and a vector of timestamps with uptime percentages
+#[derive(Serialize, Validate)]
+struct WebsiteInfo {
+    #[validate(url)]
+    url: String,
+    alias: String,
+    data: Vec<WebsiteStats>,
+}
+```
+
+To start working on our backend, we can create an initial route to add a URL to monitor. You may have noticed earlier we added the `Validate` derive trait. This allows us to validate the form data using preset rules and automatically return an error if the validation fails. In this case, we used `#[validate(url)]` - so if the string isn’t in a URL format it will automatically break:
+
+```rust
+// main.rs
+async fn create_website(
+    State(state): State<AppState>,
+    Form(new_website): Form<Website>,
+) -> Result<impl AxumIntoResponse, impl AxumIntoResponse> {
+    if new_website.validate().is_err() {
+        return Err((
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "Validation error: is your website a reachable URL?",
+        ));
+    }
+
+    sqlx::query("INSERT INTO websites (url, alias) VALUES ($1, $2)")
+        .bind(new_website.url)
+        .bind(new_website.alias)
+        .execute(&state.db)
+        .await
+        .unwrap();
+
+    Ok(Redirect::to("/"))
+```
+
+Now let’s write a route to grab all the websites we’re monitoring, as well as get a quick report on what the uptime was like in the last recorded 24 hours (filling in any gaps where required) for each website. 
+
+Like before with the monitoring tasks, we need to grab a list of all of the websites we’re currently tracking (except we will assume there are results there. If not, `askama` will handle it for us by automatically not rendering any records):
+
+```rust
+async fn get_websites(State(state): State<AppState>) -> impl AskamaIntoResponse {
+    let websites = sqlx::query_as::<_, Website>("SELECT url, alias FROM websites")
+        .fetch_all(&state.db)
+        .await
+        .unwrap();
+    // .. rest of your code
+}
+```
+
+Once this is done, we’ll then need to make a new Vector. This will hold all the website URLs (and respective aliases) as well as a list of timestamps with the uptime percentage over the last 24 hours, calculated per hour from how many HTTP requests returned with `200 OK`. We then need to check if there’s any gaps and fill them in with a `None`. This lets the person viewing the data know that no data was recorded at the time (for example if we’re either developing locally, or if the service itself had an outage and was unable to record data). 
+
+```rust
+// main.rs
+let mut logs = Vec::new();
+
+for website in websites {
+    let mut data = sqlx
+        ::query_as::<_, WebsiteStats>(
+            r#"
+            SELECT date_trunc('hour', created_at) as time,
+            CAST(COUNT(case when status = 200 then 1 end) * 100 / COUNT(*) AS int2) as uptime_pct
```

Comment:
If this service only did, let's say, 45 checks in the last hour, then won't this `COUNT(*)` be 45? Is it correct to not calculation on 60 (minutes)?

Reply:
I was a little bit unsure about this but I mainly wanted to represent the percentage of recorded uptime over the amount of time it was recorded. If we go with 60 minutes for example, the currently recorded hour may show something like 5% uptime (for example). Might need to expand on this a little bit

Reply:
Ok, sounds good for me :+1: